torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+            h = p.register_hook(lambda grad: torch.ops._spmd.tag_grad(grad))
+            tagging_hooks.append(h)
+        yield
+    finally:
+        # remove those hooks after tracing
+        for h in tagging_hooks:
+            h.remove()
+
+
+class BatchDimAnalyzer(object):
+    """
+    This class is used to analyze the batch dimension of each tensor/node in the
+    graph. We need to know the batch dimension of each tensor/node so that we know
+    exactly the sharding layout of intermediate tensors.
+
+    We possibly should use symbolic shapes to track the batch dimension, but this
+    needs to happen with dynamo, as dynamo is the only place we can mark a single
+    dimension. For now, we just use the batch dimension of the first input tensor
+    as the hint to track the batch dimension of all tensors/nodes in the graph.
+    """
+
+    def __init__(self, batch_dim: int = 0) -> None:
+        self.batch_dim = batch_dim
+
+        if batch_dim != 0:
+            # TODO: see if this make sense or not
+            raise RuntimeError("Data Parallel only supports batch dim on dimension 0!")
+
+        self.batch_dim_map: Dict[fx.Node, int] = {}
+        # batch dim size is used to track the batch dim size of the input tensor
+        self.batch_dim_size = -1
+
+    def init_batch_dim_size(self, batch_dim_size: int) -> None:
+        """
+        initialize batch dim size base on the first input batch size
+        """
+        if self.batch_dim_size != -1 and self.batch_dim_size != batch_dim_size:
+            raise RuntimeError(
+                f"batch dim size is already initialized! "
+                f"Found new batch size: {batch_dim_size} not "
+                f"matching existing batch dim size: {self.batch_dim_size}!"
+            )
+        self.batch_dim_size = batch_dim_size
+
+    def set_batch_dim(self, node: fx.Node, batch_dim: int) -> None:
+        self.batch_dim_map[node] = batch_dim
+
+    def get_batch_dim(self, node: fx.Node) -> int:
+        if node not in self.batch_dim_map:
+            raise RuntimeError(f"batch dim analysis failed on node: {node}!")
+        return self.batch_dim_map[node]
+
+    def get_batch_dim_shard_spec(
+        self, node: fx.Node, mesh: DeviceMesh, input_full_reduction: bool = False
+    ) -> Tuple[DTensorSpec, bool]:
+        """
+        simple batch dim analysis that analyze the batch dim of the node
+        and return the corresponding shard spec
+        """
+        assert self.batch_dim_size != -1, "batch dim size is not initialized!"
+
+        if node in self.batch_dim_map:
+            node_batch_dim = self.get_batch_dim(node)
+            batch_dim_shard_spec = DTensorSpec(
+                mesh=mesh, placements=[Shard(node_batch_dim)]
+            )
+            return batch_dim_shard_spec, False
+
+        shape = node.meta["val"].shape
+
+        # for reduction op that reduces over the sharded batch dim
+        # we don't generate partial, but rather, we generate shard
```

Comment:
This should work, but seems to be slightly abusing DTensor's placement definitation? E.g., if I do `sum()` and get a scalar tensor, the tensor itself actually does not have a batch dimension to shard?

Reply:
Oh yeah you are right. that actually need to be specifically handled, I used a `from_local` flag in the partitioner to indicate that this node don't really have dim and so it can skips the sharding.